lib/DataFileUtil/DataFileUtilImpl.py

```diff
@@ -245,6 +251,275 @@ def make_handle(self, shock_data, token):
     def make_ref(self, object_info):
         return str(object_info[6]) + '/' + str(object_info[0]) + \
             '/' + str(object_info[4])
+
+    def _get_staging_file_path(self, token_user, staging_file_subdir_path):
+        """
+        _get_staging_file_path: return staging area file path
+
+        directory pattern: /data/bulk/user_name/file_name
+
+        """
+        return self.STAGING_FILE_PREFIX + token_user + '/' + staging_file_subdir_path
+
+    def _download_file(self, download_type, file_url):
+        """
+        _download_file: download execution distributor
+
+        params:
+        download_type: download type for web source file
+        file_url: file URL
+
+        """
+        if download_type == 'Direct Download':
+            copy_file_path = self._download_direct_download_link(file_url)
+        elif download_type == 'DropBox':
+            copy_file_path = self._download_dropbox_link(file_url)
+        elif download_type == 'FTP':
+            copy_file_path = self._download_ftp_link(file_url)
+        elif download_type == 'Google Drive':
+            copy_file_path = self._download_google_drive_link(file_url)
+        else:
+            valid_download_types = ['Direct Download', 'FTP', 
+                                'DropBox', 'Google Drive']
+            error_msg = "[{}] download_type is invalid.\n".format(download_type)
+            error_msg += "Please use one of {}".format(valid_download_types)
+            raise ValueError(error_msg)
+
+        return copy_file_path
+
+    def _retrieve_filepath(self, file_url, download_type):
+        """
+        _retrieve_filepath: retrieve file name from download URL and return local file path
+
+        """
+        try:
+            response = requests.get(file_url, stream=True)
+            content_disposition = response.headers['content-disposition']
+        except KeyError:
+            self.log('Parsing file name directly from URL')
+            file_name = file_url.split('/')[-1]
+        else:
+            file_name = content_disposition.split('filename="')[-1].split('";')[0]
+        finally:
+            response.close()
+            
+        self.log('Retrieving file name from url: {}'.format(file_name))
+        copy_file_path = os.path.join(self.tmp, file_name)
+
+        return copy_file_path
+
+    def _download_to_file(self, file_url, copy_file_path):
+        """
+        _download_to_file: download url content to file
+
+        params:
+        file_url: direct download URL
+        copy_file_path: output file saving path
+
+        """
+
+        self.log('Connecting and downloading web source: {}'.format(
+                                                                file_url))
+        try:
+            online_file = urllib2.urlopen(file_url)
+        except urllib2.HTTPError as e:
+            raise ValueError(
+                "The server error\nURL: {}\nError code: {}".format(
```

Comment:
'The server error' doesn't seem like a full message, given it's on a single line.